src/engine/strat_engine/thinpool/filesystem.rs

```diff
@@ -130,19 +149,19 @@ impl StratFilesystem {
     #[allow(clippy::too_many_arguments)]
     pub fn snapshot(
         &self,
-        thin_pool: &ThinPoolDev,
+        thin_pool: &WithLock<ThinPoolDev>,
         snapshot_name: &str,
         snapshot_dm_name: &DmName,
         snapshot_dm_uuid: Option<&DmUuid>,
         snapshot_fs_name: &Name,
         snapshot_fs_uuid: FilesystemUuid,
         snapshot_thin_id: ThinDevId,
     ) -> StratisResult<StratFilesystem> {
-        match self.thin_dev.snapshot(
+        match self.thin_dev.lock().snapshot(
             get_dm(),
             snapshot_dm_name,
             snapshot_dm_uuid,
-            thin_pool,
+            &*thin_pool.lock(),
```

Comment:
One note about this piece of code: if this were a mutable lock acquisition, this could cause a dead lock. What are your thoughts on the design based on this @mulkieran?